### DISCUSSION TITLE
Need help updating an old project form from a verry old PKHex version

### DISCUSSION THREAD

**Discussion by ThunderSpark91** [kwsch/PKHeX#4098] [General]
Hello I am a software developer who is trying to revive a very old pokemon project that used PkeHex from more than 5 years ago. Currently I am trying to update to the newest version. Of course this will give issues with the code considering many things have changed. I am currently using PKHeX-23.10.11 to do the update.

However I have a few problems finding certain aspects or rather I have a few questions making me unsure if I am doing the right thing here. And I need to ask here to clarify.

- Deoxys Alternate form and other Pokémon alternate forms: Byte 0 is the default?
The code I have does have a function to check for the Deoxys form based on game version. This returns an in which I know convert to byte. Since the byte is now the form. So I assume when the byte is value 0, it is the default form. Is this actually true?

                default:
                    return 0;
                case GameVersion.FR: // Attack
                    return 1;
                case GameVersion.LG: // Defense
                    return 2;
                case GameVersion.E: // Speed
                    return 3;
                    

I have some code here that uses that uses the user MysteryGift class as input to get an Image back. I have an issue or two.

- Mystery Gift Pokémon Check: Is the species 0 if the mystery gift is not a Pokémon?

In the old code there is a check to see if the mystery Gift is a Pokémon or not, which was based on a boolean value in the Mystery Gift Class. **I have solved this to say if the Species is higher than 0, but I am unsure if this is the case, I am guessing here.
Is this case or not?** Also has what is this is entity boolean here? Is this what the bollean has replaced?

![afbeelding](https://github.com/kwsch/PKHeX/assets/59621887/96466917-00c4-4f24-a760-237e808ff714)

For now these are my issues but I will have many more. I will post them alter when they get resolved.
I post here to ask for help.
 
 
 ### Edit above situations except Deoxys have been resolved but there are two more issues in my code that I still need to resolve around the Mystery Gift . There also one around save data but it will be in another discussion.
 
 

-  Mysterry Gift: Was Generation Replaced with Format?

 The old fucntion I have called Image GetSprite(MysteryGift gift°. Has a special scenario  to check if the Gift is an item and takes some input in like this
 
 img = GetSprite(gift.Species, gift.Form, gift.Gender, gift.HeldItem, gift.IsEgg, gift.IsShiny, gift.Format); 
 
 The thing is Format does not exist anymore in the newer versions and from what I can tell based on this screenshots. **I suspect that Format actually now stands for Generation**. But I am only 90% sure this is the case.

![afbeelding](https://github.com/kwsch/PKHeX/assets/59621887/65f99103-490d-4533-a471-47a0df8ff7d5)

 
 
![afbeelding](https://github.com/kwsch/PKHeX/assets/59621887/f575ae4e-4664-4b93-bb65-adf6b2a1e60a)

- Legal Class Z Crystal item number:

The next following issue I have is something actually very specific that has to do with ZCrystals. You see in the code of the old project  something called the ZCrystal Dictionary is used, to get the item number of the ZCrystal and then set number that item in the application resources. This is gotten from the legal class, the thing is though that function,  no longer exists in that class.

                int item = gift.ItemID;
                if (Legal.ZCrystalDictionary.TryGetValue(item, out int value))
                    item = value;
                img = (Image)(Resources.ResourceManager.GetObject("item_" + item) ?? Resources.unknown);

So where was this old function placed and how is now called?

**Comment by kwsch** [comment]
In Generation 3, Deoxys depends on the game it needs to calculate for. Each game only has 1 form available, which has different base stats.

For mystery gifts, they can be Pokémon, and they can be items, clothing, unlocks, etc. The default for the abstract class is to just implement something and let the derived classes override the behavior.

Checking `Species != 0` is not accurate, as you should be checking `IsEntity` first. The games store their gift data as unions, so if it were an item, the offset would store the Item ID, not a Species ID. The normal flow is to check the card type first (if you only care about Pokémon, then you can just check `IsEntity`).

To reply to your edits: just reference the current source code. Generation and Context have replaced the integer-only Format for Mystery Gifts, but Format still exists for Pokémon as it's different from "original Generation".
https://github.com/kwsch/PKHeX/blob/1fe2b4f29b32c079913c05ccd0ecefceae39efa7/PKHeX.Drawing.Misc/Util/MysteryGiftSpriteUtil.cs#L34-L37

**Reply by ThunderSpark91** [reply]
Thank you for some clarification, I will edit my post later to bring attention to some other issues I have like the Z Crystal Dictionary ofh the old "legal class" and "Format" of the Mystery Gift Class. Which existed in the old code, but not anymore.

**Reply by Lusamine** [reply]
It's better to just write a reply if you're able to. Editing your post doesn't notify people looking at webhooks that more was added to the conversation. I tried changing the category to see if that gives more options; Q&A seems to be for one question, one answer.

**Reply by ThunderSpark91** [reply]
Thanks for the tip  Lusamine . Kwsch now I have another thing to find the old code It tries to pull the sprite data from the save data mainly an OrasDemo Properties and the multiplayer sprite data. Two properties that once again do not exist in the newer versions. If you can explain something for me, that would be great.

        private static Image GetSprite(SaveFile SAV)
        {
            string file = "tr_00";
            if (SAV.Generation == 6 && (SAV || SAV.ORASDEMO))
                file = $"tr_{SAV.MultiplayerSpriteID:00}";
            return Resources.ResourceManager.GetObject(file) as Image;
        }

**Reply by kwsch** [reply]
`SaveFile` is the abstract class. The derived classes implement `MultiplayerSpriteID`, not the abstract class.

I recommend searching the source code to find old properties/magic strings. Things may have moved / changed visibility, but functionality is never removed.
